File: packages/emitter/testcases/common/classes/generic-inheritance/ConcreteExtends.ts
import { int } from "@tsonic/core/types.js";

// Generic base class
export class Container<T> {
  value: T;
  constructor(value: T) {
    this.value = value;
  }
  getValue(): T {
    return this.value;
  }
}

// Non-generic extending generic with concrete type
export class IntContainer extends Container<int> {
  double(): int {
    return (this.getValue() * 2) as int;
  }
}

export class StringContainer extends Container<string> {
  getLength(): int {
    return this.getValue().length as int;
  }
}

File: packages/targets/csharp/emitter/testcases/common/expected/classes/generic-inheritance/ConcreteExtends.cs
namespace TestCases.common.classes.genericinheritance
{
    public class Container<T>
    {
        public T value { get; set; }

        public Container(T value)
        {
            this.value = value;
        }

        public T getValue()
        {
            return this.value;
        }
    }

    public class IntContainer : Container<int>
    {
        public IntContainer(int value) : base(value)
        {

        }

        public int @double()
        {
            return this.getValue() * 2;
        }
    }

    public class StringContainer : Container<string>
    {
        public StringContainer(string value) : base(value)
        {

        }

        public int getLength()
        {
            return this.getValue().Length;
        }
    }
}